tokio/src/macros/join.rs

```diff
@@ -74,6 +74,12 @@ macro_rules! join {
         // the requirement of `Pin::new_unchecked` called below.
         let mut futures = ( $( maybe_done($e), )* );
 
+        // This assignment makes sure that the `poll_fn` closure only has a
+        // reference to the futures, instead of taking ownership of them. This
+        // mitigates the issue described in
+        // <https://internals.rust-lang.org/t/surprising-soundness-trouble-around-pollfn/17484>
```

Comment:
Specifically, this means that `futures` in pinned in the caller stack frame, not in the closure. If tokio has a macro like `std::pin::pin!`, that could also be used here and would even avoid some unsafe calls to `Pin::new_unchecked`.

Reply:
We have such a macro [right here](https://docs.rs/tokio/latest/tokio/macro.pin.html), but we can't use it since `futures` is actually a tuple of futures, and I don't think std provides any way to pin-project over tuples.

Reply:
Ah, yeah makes sense. If the comment explained that, I think it would help readers of this code.

Reply:
I've added comments.

Reply:
the comments are helpful, thanks! one nit is that i wonder if it's worth stating explicitly in the comment that this is similar to what the `tokio::pin!` macro would do?